File: libs/songbook/src/main/kotlin/io/patterueldev/songbook/songdetails/SongDetails.kt
package io.patterueldev.songbook.songdetails

interface SongDetails {
    val id: String
    val source: String
    val title: String
    val artist: String
    val language: String
    val isOffVocal: Boolean
    val videoHasLyrics: Boolean
    val duration: Int
    val genres: List<String>
    val tags: List<String>
    val metadata: Map<String, String>
    val thumbnailPath: String
    val wasReserved: Boolean
    val currentPlaying: Boolean
    val lyrics: String?
    val addedBy: String
    val addedAtSession: String
    val lastUpdatedBy: String
    val isCorrupted: Boolean
}

File: client/packages/commonds/lib/converters/song_details.dart
part of '../commonds.dart';

extension APISongDetailsExtension on APISongDetails {
  SongDetails toSongDetails(SingalongConfiguration configuration) {
    final song = this;
    return SongDetails(
      id: song.id,
      source: song.source,
      title: song.title,
      artist: song.artist,
      language: song.language,
      isOffVocal: song.isOffVocal,
      videoHasLyrics: song.videoHasLyrics,
      duration: song.duration,
      genres: song.genres,
      tags: song.tags,
      metadata: song.metadata,
      thumbnailURL:
          configuration.buildResourceURL(song.thumbnailPath).toString(),
      currentPlaying: song.currentPlaying,
      lyrics: song.lyrics,
      addedBy: song.addedBy,
      addedAtSession: song.addedAtSession,
      lastUpdatedBy: song.lastUpdatedBy,
      isCorrupted: song.isCorrupted,
    );
  }
}
